In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression

# GETTING THE DATA

In [2]:
df = pd.read_csv("framingham.csv", low_memory=False)

y= df['TenYearCHD']
X = df.drop('TenYearCHD', axis=1)
X = X.drop('education', axis=1)

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# LOGISTIC POLINOMIAL REGRESSION

In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline

accuracys = []

for degree in range(1,5):
    lr_model = make_pipeline(
        PolynomialFeatures(degree),
        LogisticRegression(solver='lbfgs', max_iter=5000)
    )

    # Entrenar el modelo con los datos de entrenamiento
    lr_model.fit(X_train_imputed, y_train)
    y_pred = lr_model.predict(X_test_imputed)
    accuracys.append(accuracy_score(y_test, y_pred))

c:\users\juanc\appdata\local\programs\python\python38\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\users\juanc\appdata\local\programs\python\python38\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stab

In [12]:
biggest= 0
indx = -1
for i in range(0,len(accuracys)):
    if accuracys[i] > biggest:
        biggest = accuracys[i]
        indx = i
print("best degree found: " + str(indx+1))
print("accuracy of degree: " + str(biggest))

best degree found: 2
accuracy of degree: 0.8455188679245284


# GRADIENT DESCENT

In [4]:
from sklearn.impute import SimpleImputer

# Crea un objeto SimpleImputer que reemplaza los valores faltantes con la media
imputer = SimpleImputer(strategy='mean')

# Ajusta el imputer a los datos de entrenamiento y transforma los datos de entrenamiento
X_train_imputed = imputer.fit_transform(X_train)

# Usa el mismo imputer para transformar los datos de prueba
X_test_imputed = imputer.transform(X_test)

In [5]:
lr = LogisticRegression(solver='sag', max_iter=10000)
lr.fit(X_train_imputed, y_train)
y_pred = lr.predict(X_test_imputed)


In [6]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)
#precision_score(y_test, y_pred)
#recall_score(y_test, y_pred)

0.8408018867924528

# CROSS VALIDATION

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score


# Crear una instancia del modelo de regresión logística
lr_model = LogisticRegression(solver='lbfgs',  max_iter= 5000)

# Realizar la validación cruzada con 5 particiones
scores = cross_val_score(lr_model, X_train_imputed, y_train, cv=5)

# Imprimir los resultados de la validación cruzada
print('Accuracy: %0.2f (+/- %0.2f)' % (scores.mean(), scores.std() * 2))

Accuracy: 0.86 (+/- 0.01)


# HALLAZGOS

Luego de haber realizado varios modelos encontramos que el mejor modelo es utilizar el regresión logística con un gradiente de 2. Con el cual llegamos a un 84% de acertación. Sin embargo, cabe resaltar, que debido a los pocos datos que encontramos de casos positivos, la cantidad de casos negativos pero que son realmente positivos es alta, por lo que se recomienda implementar métodos para nivelar la información o quizas intentar otro modelo que nos permita aumentar la cantidad de verdaderos positivos, pues al final es lo más imn